[fetch_depth_layer] Fix normals_estimator_ no match for ‘operator=’ error

josiahls opened the pull request:
Fix how `normals_estimator_ cv::Ptr` was being set.

Set the `normals_estimator_` using `.reset()` method instead of `operator=`. 

- The [opencv 4.0.0 reset()](https://docs.opencv.org/master/d0/de7/structcv_1_1Ptr.html#a8100350b806a4e664cbec3445b8d30a0) expects a `Y *` which is what the `new `keyword will be returning. 
- The [opencv 4.0.0 operator=](https://docs.opencv.org/master/d0/de7/structcv_1_1Ptr.html#a468377d212f11a2501dcb3fab37955cd) expects a `const cv:Ptr<T> &` which is not what the `new` keyword is returning.

The result is changing:
- normals_estimator_ = new RgbdNormals( ...
To:
- normals_estimator_.reset(new RgbdNormals( ...

Note, as of 7/22/2018 #81 has been referenced to get this to work, and modified the files specified in that issue along with my one commit.

Extra Contextual Info:
This commit fits closer to how smart pointers are supposed to be used in Boost C++11 libraries. The [opencv Ptr](https://docs.opencv.org/3.4/d0/de7/structcv_1_1Ptr.html#details) is similar to [boost::shared_ptr](https://theboostcpplibraries.com/boost.smartpointers-shared-ownership). Using `reset()` for pointing a smart pointer to a new instance seems to be the "more" correct way to use these pointers. Using the `operator=` seems unusual. 

The error below would occur without this commit:
```
~$ pkg-config --modversion opencv
4.0.0
~/PycharmProjects/fetch_robot_ws$ catkin_make
-- The C compiler identification is GNU 7.3.0
-- The CXX compiler identification is GNU 7.3.0
-- Boost version: 1.65.1
...
[ 97%] Building CXX object navigation/move_base/CMakeFiles/move_base.dir/src/move_base.cpp.o
/home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/src/depth_layer.cpp: In member function ‘void costmap_2d::FetchDepthLayer::depthImageCallback(const ConstPtr&)’:
/home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/src/depth_layer.cpp:239:28: error: no match for ‘operator=’ (operand types are ‘cv::Ptr<cv::rgbd::RgbdNormals>’ and ‘cv::rgbd::RgbdNormals*’)
       normals_estimator_ = new RgbdNormals(cv_ptr->image.rows,cv_ptr->image.cols,  cv_ptr->image.depth(),K_);
                            ^~~~
In file included from /usr/local/include/opencv2/core/cvstd.hpp:1033:0,
                 from /usr/local/include/opencv2/core/base.hpp:58,
                 from /usr/local/include/opencv2/core.hpp:54,
                 from /usr/local/include/opencv2/core/core.hpp:48,
                 from /opt/ros/melodic/include/cv_bridge/cv_bridge.h:43,
                 from /home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/include/fetch_depth_layer/depth_layer.h:37,
                 from /home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/src/depth_layer.cpp:32:
/usr/local/include/opencv2/core/ptr.inl.hpp:160:9: note: candidate: cv::Ptr<T>& cv::Ptr<T>::operator=(const cv::Ptr<T>&) [with T = cv::rgbd::RgbdNormals]
 Ptr<T>& Ptr<T>::operator = (const Ptr<T>& o)
         ^~~~~~
```

josiahls committed:
Update depth_layer no match for ‘operator=’ error

I recently checked out this branch and attempted to run catkin_make. There was an issue with how the cv::Ptr was being set. 

This is fixed by setting the normals_estimator_ pointer similarly to how the plane_estimator_ pointer is being set. The package can now compile. I am confused why I needed to change this, but I am curious to learn why.

The error below would occur:
[ 97%] Building CXX object navigation/move_base/CMakeFiles/move_base.dir/src/move_base.cpp.o
/home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/src/depth_layer.cpp: In member function ‘void costmap_2d::FetchDepthLayer::depthImageCallback(const ConstPtr&)’:
/home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/src/depth_layer.cpp:239:28: error: no match for ‘operator=’ (operand types are ‘cv::Ptr<cv::rgbd::RgbdNormals>’ and ‘cv::rgbd::RgbdNormals*’)
       normals_estimator_ = new RgbdNormals(cv_ptr->image.rows,cv_ptr->image.cols,  cv_ptr->image.depth(),K_);
                            ^~~~
In file included from /usr/local/include/opencv2/core/cvstd.hpp:1033:0,
                 from /usr/local/include/opencv2/core/base.hpp:58,
                 from /usr/local/include/opencv2/core.hpp:54,
                 from /usr/local/include/opencv2/core/core.hpp:48,
                 from /opt/ros/melodic/include/cv_bridge/cv_bridge.h:43,
                 from /home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/include/fetch_depth_layer/depth_layer.h:37,
                 from /home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/src/depth_layer.cpp:32:
/usr/local/include/opencv2/core/ptr.inl.hpp:160:9: note: candidate: cv::Ptr<T>& cv::Ptr<T>::operator=(const cv::Ptr<T>&) [with T = cv::rgbd::RgbdNormals]
 Ptr<T>& Ptr<T>::operator = (const Ptr<T>& o)
         ^~~~~~
/usr/local/include/opencv2/core/ptr.inl.hpp:160:9: note:   no known conversion for argument 1 from ‘cv::rgbd::RgbdNormals*’ to ‘const cv::Ptr<cv::rgbd::RgbdNormals>&’
In file included from /usr/local/include/opencv2/core/base.hpp:58:0,
                 from /usr/local/include/opencv2/core.hpp:54,
                 from /usr/local/include/opencv2/core/core.hpp:48,
                 from /opt/ros/melodic/include/cv_bridge/cv_bridge.h:43,
                 from /home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/include/fetch_depth_layer/depth_layer.h:37,
                 from /home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/src/depth_layer.cpp:32:
/usr/local/include/opencv2/core/cvstd.hpp:333:10: note: candidate: template<class Y> cv::Ptr<T>& cv::Ptr<T>::operator=(const cv::Ptr<Y>&) [with Y = Y; T = cv::rgbd::RgbdNormals]
     Ptr& operator = (const Ptr<Y>& o);
          ^~~~~~~~
/usr/local/include/opencv2/core/cvstd.hpp:333:10: note:   template argument deduction/substitution failed:
/home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/src/depth_layer.cpp:239:28: note:   mismatched types ‘const cv::Ptr<T>’ and ‘cv::rgbd::RgbdNormals*’
       normals_estimator_ = temp;
                            ^~~~
In file included from /usr/local/include/opencv2/core/cvstd.hpp:1033:0,
                 from /usr/local/include/opencv2/core/base.hpp:58,
                 from /usr/local/include/opencv2/core.hpp:54,
                 from /usr/local/include/opencv2/core/core.hpp:48,
                 from /opt/ros/melodic/include/cv_bridge/cv_bridge.h:43,
                 from /home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/include/fetch_depth_layer/depth_layer.h:37,
                 from /home/josiah/PycharmProjects/fetch_robot_ws/src/fetch_ros/fetch_depth_layer/src/depth_layer.cpp:32:
/usr/local/include/opencv2/core/ptr.inl.hpp:263:9: note: candidate: cv::Ptr<T>& cv::Ptr<T>::operator=(cv::Ptr<T>&&) [with T = cv::rgbd::RgbdNormals]
 Ptr<T>& Ptr<T>::operator = (Ptr<T>&& o)
         ^~~~~~
/usr/local/include/opencv2/core/ptr.inl.hpp:263:9: note:   no known conversion for argument 1 from ‘cv::rgbd::RgbdNormals*’ to ‘cv::Ptr<cv::rgbd::RgbdNormals>&&’
fetch_ros/fetch_depth_layer/CMakeFiles/fetch_depth_layer.dir/build.make:62: recipe for target 'fetch_ros/fetch_depth_layer/CMakeFiles/fetch_depth_layer.dir/src/depth_layer.cpp.o' failed
make[2]: *** [fetch_ros/fetch_depth_layer/CMakeFiles/fetch_depth_layer.dir/src/depth_layer.cpp.o] Error 1
CMakeFiles/Makefile2:10133: recipe for target 'fetch_ros/fetch_depth_layer/CMakeFiles/fetch_depth_layer.dir/all' failed
make[1]: *** [fetch_ros/fetch_depth_layer/CMakeFiles/fetch_depth_layer.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 98%] Linking CXX shared library /home/josiah/PycharmProjects/fetch_robot_ws/devel/lib/libmove_base.so
[ 98%] Built target move_base
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
Invoking "make -j4 -l4" failed

moriarty commented:
It's strange that I can't reproduce this compile error.
I recently cleaned up my work-space when I found #80 

Now that the ros-planning/navigation#755 is merged, we can merge in #81.

First, I'll delete my work-space and try to reproduce.

Can you provide some info on how you installed OpenCV? And which version it is?

Also, before accepting this could you clean up the commit message a bit?
the first bit is fine it doesn't need _so much_ compile error output.
https://chris.beams.io/posts/git-commit/

josiahls commented:
I'll clean it up tonight and try it out on a different computer. I will
also verify the opencv version.

Thanks,
Josiah L.

On Sun, Jul 22, 2018, 7:08 PM Alex Moriarty <notifications@github.com>
wrote:

> It's strange that I can't reproduce this compile error.
> I recently cleaned up my work-space when I found #80
> <https://github.com/fetchrobotics/fetch_ros/pull/80>
>
> Now that the ros-planning/navigation#755
> <https://github.com/ros-planning/navigation/pull/755> is merged, we can
> merge in #81 <https://github.com/fetchrobotics/fetch_ros/pull/81>.
>
> First, I'll delete my work-space and try to reproduce.
>
> Can you provide some info on how you installed OpenCV? And which version
> it is?
>
> Also, before accepting this could you clean up the commit message a bit?
> the first bit is fine it doesn't need *so much* compile error output.
> https://chris.beams.io/posts/git-commit/
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/fetchrobotics/fetch_ros/pull/82#issuecomment-406904390>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATAdc7sSY9UUOpOD6ReVSCyAH1VGPR31ks5uJQX3gaJpZM4VaKYX>
> .
>

moriarty commented:
Thanks.

I cleaned up my workspace 3 times (nuclear option: rm -rf the build, devel, logs and install directories) and tried with each: **catkin build**, **catkin_make** and **catkin_make_isolated** and wasn't able to reproduce.

moriarty commented:
The plane_estimator_ needs the .reset and calls to each setter, because there is no appropriate constructor.

The normals_estimator_ constructor should work fine...

If you can let us know more details then I think the cleaned up commit in a8fe874 can be merged.

josiahls committed:
Cleaned up normals_estimator_ Ptr setting.

Cleaned the code to where normals_estimator_ is being set. 

Needs to use the reset method for setting the pointer.

josiahls commented:
I am using opencv 4.0.0. I updated the above commit message to be a little cleaner, and concise.

I modified the commit to address the primary cause of the error. I believe the `normals_estimator_ and plane_estimator_ both` need to use the `reset()` method since they are both smart pointers.

Opencv was installed via  [git and cmake](https://docs.opencv.org/3.4.1/d2/de6/tutorial_py_setup_in_ubuntu.html)

The way that opencv 4.0.0 docs, and current boost smart pointer docs specify smart pointer usage is to use the `reset()` method. I do not understand why the compiler error does not happen on your system, however my system is completely wiped, using the newest opencv , ubuntu 18.04.

I have compiled the older version of this file fetch_ros indigo without this compiler error with an older version of opencv. I do not remember what the versions where unfortunately.  
```
normals_estimator_.reset(new RgbdNormals(cv_ptr->image.rows,
                                               cv_ptr->image.cols,  
                                               cv_ptr->image.depth(),
                                               K_));
```

josiahls commented:
Ok I think I know why this compiler error happens. You need to have the newest version of Opencv:

Compare [opencv 4.0.0](https://docs.opencv.org/master/d1/d5b/classcv_1_1rgbd_1_1RgbdNormals.html)
Compare [opencv 3.0.0](https://docs.opencv.org/3.0.0/d1/d5b/classcv_1_1rgbd_1_1RgbdNormals.html)

Version 4.0.0 uses: `create()` for returning a `Ptr<RgbdNormals>`
Version 3.0.0 not sure. The `create` method is not implemented here. The class uses `Additional Inherited Members` which return `Ptr` but otherwise I am not sure how the `new RgbdNormals` converts to a `Ptr`.

What this tells me is that as of 4.0.0, you need to call `create()` if you want to get a `Ptr`. I still stand by just using `reset()` however. Let me know if that makes sense.

So...

If you update to opencv 4.0.0, you will get the compiler error. That is why wiping the workspace didn't make a difference.

josiahls commented:
I am using opencv 4.0.0. I updated the above commit message to be a little cleaner, and concise.

I modified the commit to address the primary cause of the error. I believe the normals_estimator_ and plane_estimator_ both need to use the `reset()` method since they are both smart pointers.

moriarty commented:
Okay OpenCV 4.0.0 is still in pre-release. So that's why you have the error and we haven't seen it.

moriarty commented on fetch_depth_layer/src/depth_layer.cpp:
```diff
@@ -232,10 +232,10 @@ void FetchDepthLayer::depthImageCallback(
-                                           cv_ptr->image.depth(),
-                                           K_);
+      normals_estimator_.reset(new RgbdNormals(cv_ptr->image.rows,
+                                               cv_ptr->image.cols,  
```
```suggestion
                                               cv_ptr->image.cols,
```

moriarty replied:
remove extra whitespace and then this can be merged via the GitHub `squash & merge` feature to cleanup the commit message from the first commit

moriarty reviewed (changes requested):
sorry for the long delay. just a little whitespace fix and then I think we can merge this with a squash and merge to keep the commits tidy.
